src/PlaneWaveBasis.jl

```diff
@@ -177,23 +186,47 @@ build_kpoints(basis::PlaneWaveBasis, kcoords) =
         symmetries = vcat(ksymops...)
     end
 
-    n_spin = model.n_spin_components
-    @assert n_spin in (1, 2)  # For 1 we are all set
-    n_spin == 2 && (ksymops = vcat(ksymops, ksymops))
+    # At this point, select the subset of kpoints to be computed by this processor
+    mpi_comm = MPI.COMM_WORLD
+    nprocs = mpi_nprocs(mpi_comm)
+    # Right now we split only the kcoords: both spin channels have to be handled by the same process
+    nkpt = length(kcoords)
+    if nprocs > nkpt
+        @warn "No point in trying to parallelize $nkpt kpoints over $nprocs processes; falling back to sequential"
+        # supporting this would require fixing a bunch of "reducing over empty collections not allowed" errors
+        mpi_comm = MPI.COMM_SELF # everybody talks with himself now: reduces are no-ops
+        krange_thisproc = 1:nkpt
+    else
+        krange_thisproc = mpi_split_work(mpi_comm, nkpt)  # get the slice of 1:nkpt handled by this process
+        @assert mpi_sum(length(krange_thisproc), mpi_comm) == nkpt
+    end
+
+    kcoords = kcoords[krange_thisproc]
+    ksymops = ksymops[krange_thisproc]
+
+    # Build kpoint-specific basis. Notice that this also builds index mapping from the k-point-specific basis
+    # to the global basis and thus the fft_size needs to be final at this point.
+    kpoints = build_kpoints(model, fft_size, kcoords, Ecut; variational=variational)
+
+    # kpoints is now possibly twice the size of kcoords. Double it for spin
+    if model.n_spin_components == 2
+        ksymops = vcat(ksymops, ksymops)
+        krange_thisproc = vcat(krange_thisproc, nkpt .+ krange_thisproc)
+    end
```

Comment:
This function keeps getting longer. Any chance on refactoring something into another function?

Reply:
I know, it's getting out of hand... But really it's one of the rare cases where long functions are OK: it's just initializing a lot of things in sequence. We can split it up but it reduce readability rather than the reverse

Reply:
Hmm. I'm not convinced ... for Model there turned out to be a lot of simplification possible last time I worked on the initialiser.  Can you make a TODO. I'll take a look next time I have to work on PlaneWaveBasis.

Reply:
I agree we should simplify it, especially the variational bit